docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,205 @@
+pipeline state.  The actual shader code does not matter.  These entries must not
+be present or
+set to 0 in unlinked shaders.  This removes the dependency of unlinked shaders on
+those parts of the pipeline state.
+
+## Epilogues and prologues
+
+Some changes to the text section are too large to be handled by relocations.
+The two main examples are the vertex attribute input formates and the color
+export formates.  These are handled by generating a prologue or epilogue for the
+unlinked shaders.  The prologue and epilogue are also referred to as glue
+shaders, and are implemented by extending the `GlueShader` class in
+[GlueShader.h](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/elfLinker/GlueShader.h).
+
+When generating an unlinked vertex shader, the entry point will have the name
+`_amdgpu_vs_main_fetchless`.  The inputs to this shader will be all of the
+regular inputs to a vertex shader plus all of the vertex input attributes in
+registers.  The PAL metadata will contain a table called `.vertexInputs` that
+will pass along information from the original llvm-ir that would otherwise be
+lost, but is needed to load the vertex inputs correctly.  When linking, a
+prologue is prepended to the unlinked vertex shader that uses the pipeline
+state and the `.vertexInputs` metadata to load all of the vertex input
+attributes into registers. See the `FetchShader` class in
+[GlueShader.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/elfLinker/GlueShader.cpp)
+for the generation of the prologue. To see how the unlinked shader handles
+vertex inputs see the `LowerVertexFetch` pass in
+[VertexFetch.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/patch/VertexFetch.cpp).
+
+Fragment shaders will require a prologue to export its outputs.  The unlinked fragment
+shader will have the standard name for the fragment shader.  Instead of
+exporting the outputs, they are left in registers at the end of the shader,
+where there is no `s_endpgm`. Extra information is add to the PAL metadata in
+the `.colorExport` table.  During linking this table and the pipeline state are
+used to generate all of the color export instructions and they are appended to
+the end of the unlinked shader. See the `ColorExportShader` class in
+[GlueShader.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/elfLinker/GlueShader.cpp).
+for the generation of the epilogue.
+To see how the unlinked shader handles
+color export see the `LowerFragColorExport` pass in
+[FragColorExport.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/patch/FragColorExport.cpp).
+
+These epilogues and prologues mean that unlinked shader do not have to depend on
+the format of the vertex input attributes and color exports.
+
+## User data nodes
+
+The user data nodes are handled differently in unlinked shaders. The
+differences are best described with examples.  Suppose you want to load the
+descriptor for descriptor set 2 at binding 1.
```

Comment:
Could you paste a few examples of data nodes here? Then 1. and 2. could refer to them, so it's easier to see what it means for a data node to be top-level or not.

Reply:
I've added diagram to explain the cases for a standard compile, and I've added a new test that I point to for the relocatable case.  I've also rewritten the section.  Let me know if it is clearer now.